articles/testing-documentation-snippets/article.md

```diff
@@ -0,0 +1,173 @@
+
+Here we are, we have a working markdown templating system with real source code files as  snippets examples, able to generate HTML.
+
+![schema1](./images/schema1.png)
+
+## Executing these snippets
+
+By definition, snippets are just a piece of code and they aren't often able to be played without a context (other pieces of code). We imagined a template system in which snippets content will be injected.
+
+Here, an example of a snippet in javascript
+
+```js
+//Connects to Kuzzle
+kuzzle
+  .connect()
+  .then(() => {
+    console.log('Successfully connected');
+  })
+  .catch(error => {
+    console.error(error.message);
+  });
+```
+
+And here a code template
+
+```js
+// load the Kuzzle SDK module
+const { Kuzzle } = require('kuzzle-sdk');
+
+// instantiate a Kuzzle client
+const kuzzle = new Kuzzle('websocket', {
+  host: 'kuzzle',
+  autoReconnect: false
+});
+
+// snippet will be injected here
+[snippet-code]
+
+```
+
+The tag **[snippet-code]** in the template will be replaced by the snippet and will form a full context file which can be executed.
+
+![schema2](./images/schema2.png)
+
+
+At this point, we have many source files, in many languages. First version of this new documentation system can support snippets test for Javascript, Java, C++ and Go. Some languages are interpreted, others compiled. All of them need their dedicated toolchain !
+
+![use docker](./images/use-docker.jpg)
+
+First thing I do is to create a DockerFile for each language implemented with all the toolchain to compile/interpret generated file. Also, we decided to add a linter in each container for each language to be sure to have a correct syntax and minimize source of errors (you can see an exemple of a dockerfile [here](https://hub.docker.com/r/kuzzleio/documentation-v2/~/dockerfile/)).
+
+At this point, we had some docker containers able to run code for their dedicated language and things seem to work perfectly!
+
+![schema3](./images/schema3.png)
+
+
+## Implementing test system
+
+![doge](./images/doge.png)
+
+Next task was to implement the test system properly speaking. How to test all of these languages with one test definition? Finally that's simple, all languages can write in the standard output. Javascript has his `console.log`, C++ his `std::cout`, Go his `fmt.Println` etc, etc ... With the help of my teammates we decided to use **Expect** to check if the tests are passed or not.
+
+> *From wikipedia :* Expect is an automation and non-regression testing tool written by Don Libes as an extension to the Tcl scripting language to test interactive applications like telnet, ftp, passwd, fsck, rlogin, ssh, or others.
+
+Running tests is simple, we give to expect a command to execute, also a string expected and it can do assertion on the result of the command.
```

Comment:
Running tests is simple, tell Expect the command to execute and our assertions on the resulting standard output.